### PR TITLE
feat: set prevrandao and block gaslimit

### PR DESCRIPTION
**Pull request by enitrat**
# Pull Request type

Time spent on this PR: 0.3d

Resolves: #{Issue number}

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: crates/ef-testing/src/evm_sequencer/evm_state/v0.rs [hunk 1]
```diff
@@ -38,15 +43,43 @@ impl Evm for KakarotSequencer {
...
+            .set_storage_at(kakarot_address, basefee_address, StarkFelt::from(low_fee));
         self.state_mut().set_storage_at(
             kakarot_address,
-            next_storage_key(&base_address)?,
+            next_storage_key(&basefee_address)?,
             StarkFelt::from(high_fee),
         );
 
+        // Set the previous randao.
+        let prev_randao_low = prev_randao & U256::from(u128::MAX);
```
Comment (greged93) [inline crates/ef-testing/src/evm_sequencer/evm_state/v0.rs:56]: There should be a helper function split_u256 in the ef tests I think (not on pc so can't check)

File: crates/ef-testing/src/evm_sequencer/evm_state/v0.rs [hunk 2]
```diff
@@ -38,15 +43,43 @@ impl Evm for KakarotSequencer {
...
+            StarkFelt::from(prev_randao_low),
+        );
+        self.state_mut().set_storage_at(
+            kakarot_address,
+            next_storage_key(&prev_randao_address)?,
+            StarkFelt::from(prev_randao_high),
+        );
+
+        // Set the block gas limit, using the 128 lower bits.
+        let block_gaslimit_low = block_gaslimit & U256::from(u128::MAX);
```
Comment (greged93) [inline crates/ef-testing/src/evm_sequencer/evm_state/v0.rs:74]: Same I think there is a helper

File: crates/ef-testing/src/evm_sequencer/evm_state/mod.rs [hunk 3]
```diff
@@ -17,7 +17,12 @@ use super::account::KakarotAccount;
 pub trait Evm {
     // TODO enforce using a marker type that you can only proceed
     // with execution if the state is initialized.
-    fn setup_state(&mut self, _base_fee: U256) -> StateResult<()> {
+    fn setup_state(
+        &mut self,
+        _base_fee: U256,
+        _prev_randao: U256,
+        _block_gaslimit: U256,
```
Comment (greged93) [inline crates/ef-testing/src/evm_sequencer/evm_state/mod.rs:24]: block_gas_limit

File: crates/ef-testing/src/evm_sequencer/evm_state/v0.rs [hunk 4]
```diff
@@ -21,7 +21,12 @@ use crate::starknet_storage;
 
 impl Evm for KakarotSequencer {
     /// Sets up the evm state (coinbase, block number, etc.)
-    fn setup_state(&mut self, base_fee: U256) -> StateResult<()> {
+    fn setup_state(
+        &mut self,
+        base_fee: U256,
+        prev_randao: U256,
+        block_gaslimit: U256,
```
Comment (greged93) [inline crates/ef-testing/src/evm_sequencer/evm_state/v0.rs:28]: block_gas_limit

File: crates/ef-testing/src/evm_sequencer/evm_state/v0.rs [hunk 5]
```diff
@@ -33,20 +38,40 @@ impl Evm for KakarotSequencer {
...
+        );
+        self.state_mut().set_storage_at(
+            kakarot_address,
+            next_storage_key(&prev_randao_address)?,
+            StarkFelt::from(high_prev_randao),
+        );
+
+        // Set the block gas limit, considering it fits in a felt.
+        let block_gaslimit_as_hexstring = block_gaslimit.to_be_bytes();
+        let block_gaslimit = StarkFelt::new(block_gaslimit_as_hexstring).unwrap();
```
Comment (greged93) [inline crates/ef-testing/src/evm_sequencer/evm_state/v0.rs:67]:
```suggestion
        let [block_gas_limit, _] = split_u256(block_gas_limit);
        let block_gaslimit = StarkFelt::from(block_gas_limit);
```

Comment (enitrat) [reply crates/ef-testing/src/evm_sequencer/evm_state/v0.rs:67]: I don't want it as u256, and I want to keep eventual bytes above the 128th one

Comment (enitrat) [reply crates/ef-testing/src/evm_sequencer/evm_state/v0.rs:67]: even though in practice it's very unlikely to be > 128 bits I think felt from bytes is a correct approach here

Comment (greged93) [reply crates/ef-testing/src/evm_sequencer/evm_state/v0.rs]: either it's u256 and then we need to store it in two storage slots, or it's u128 and then we split it. There isn't really any different values in the EVM

Comment (greged93) [reply crates/ef-testing/src/evm_sequencer/evm_state/v0.rs]: fyi in the `SealedHeader` type from Reth, they use `u64`.

Comment (enitrat) [reply crates/ef-testing/src/evm_sequencer/evm_state/v0.rs]: ok 👍

**Review by Eikix — APPROVED** [review]

**Pull request merged**